### PR TITLE
Support building of prefix if it isn't found

### PR DESCRIPTION
**Pull request by terjekv**

### CODE REVIEW

File: ansible/playbooks/roles/compatibility_layer/defaults/main.yml [hunk 1]
```diff
@@ -9,7 +9,17 @@ cvmfs_publish_transaction: False
...
 cvmfs_repository: pilot.eessi-hpc.org
 
-gentoo_prefix_path: /cvmfs/pilot.eessi-hpc.org/2020.10/compat/x86_64
+gentoo_prefix_path: /cvmfs/pilot.eessi-hpc.org/2020.10/compat/{{ eessi_host_arch }}
+
+# How to build the prefix
+prefix_singularity_command: "singularity run -B{{ gentoo_prefix_path }}:{{ gentoo_prefix_path }}"
+prefix_source: "docker://eessi/bootstrap-prefix:centos8-{{ eessi_host_arch }}"
+prefix_source_options: "{{ gentoo_prefix_path }} noninteractive"
+prefix_install: "{{ prefix_singularity_command}} {{ prefix_source }} {{ prefix_source_options }}"
```
Comment (bedroge) [inline ansible/playbooks/roles/compatibility_layer/defaults/main.yml:18]:
```suggestion
prefix_install: "{{ prefix_singularity_command }} {{ prefix_source }} {{ prefix_source_options }}"
```
Yeah, I like spaces :-)

File: ansible/playbooks/roles/compatibility_layer/defaults/main.yml [hunk 2]
```diff
@@ -9,7 +9,17 @@ cvmfs_publish_transaction: False
...
+
+# Logging
+eessi_log_dir: "/tmp/eessi-logs/"
+prefix_build_log: "{{ eessi_log_dir }}prefix-build.log"
```
Comment (bedroge) [inline ansible/playbooks/roles/compatibility_layer/defaults/main.yml:22]:
```suggestion
eessi_log_dir: "/tmp/eessi-logs"
prefix_build_log: "{{ eessi_log_dir }}/prefix-build.log"
```
I think this looks a bit cleaner.

File: ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
+# Install a specified list of sets and packages.
```
Comment (bedroge) [inline ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml:1]:
```suggestion
# Install Gentoo Prefix.
```

File: ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+- name: "Install EPEL"
+  yum:
+    name:
+      - https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
+    disable_gpg_check: yes
+    state: present
+  tags:
+    - build_prefix
+
+- name: "Install other requirements (singularity et al)"
```
Comment (bedroge) [inline ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml:12]:
```suggestion
- name: "Install other requirements (Singularity et al.)"
```

File: ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    mode: "0755"
+  with_items:
+    - "{{ gentoo_prefix_path }}"
+    - "{{ eessi_log_dir }}"
+  tags:
+    - build_prefix
+
+- name: "Give {{ username_on_host.stdout }} recursive user and group ownership of /cvmfs"
+  file:
+    dest: "cvmfs"
```
Comment (bedroge) [inline ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml:44]:
```suggestion
    dest: "/cvmfs"
```

**Review by bedroge — CHANGES_REQUESTED** [review]

File: ansible/playbooks/roles/compatibility_layer/defaults/main.yml [hunk 6]
```diff
@@ -9,7 +9,17 @@ cvmfs_publish_transaction: False
 cvmfs_abort_transaction_on_failures: False
 cvmfs_repository: pilot.eessi-hpc.org
 
-gentoo_prefix_path: /cvmfs/pilot.eessi-hpc.org/2020.10/compat/x86_64
+gentoo_prefix_path: /cvmfs/pilot.eessi-hpc.org/2020.10/compat/{{ eessi_host_arch }}
```
[the remark this answers is not in the document]
Comment (peterstol) [reply ansible/playbooks/roles/compatibility_layer/defaults/main.yml:12]: eessi_version might be ambiguous with the software layer

[the remark this answers is not in the document]
Comment (terjekv) [reply ansible/playbooks/roles/compatibility_layer/defaults/main.yml:12]: @peterstol You're thinking we may install the software layer on a non-current overlay version?

File: ansible/playbooks/README.md [hunk 7]
```diff
@@ -49,11 +49,11 @@ Before running the playbook, make sure the following settings are correct, and o
...
+ansible-playbook -i hosts -b install.yml
 ```
-The `-K` option will ask for your sudo password, and you have to supply a valid hosts file (here named `hosts`).
+The `-b` option will assume you can become root without a sudo password, and you have to supply a valid hosts file (here named `hosts`).
 By default, the playbook will only run on the host listed in the `cvmfsstratum0servers` section of the supplied `hosts` file. So, your `hosts` file should at least have:
 ```
 [cvmfsstratum0servers]
-ip-or-hostname-of-your-stratum0
+ip-or-hostname-of-your-stratum0 eessi_host_arch=x86_64
 ```
```
[the remark this answers is not in the document]
Comment (terjekv) [reply ansible/playbooks/README.md:59]: Added, thanks!

File: ansible/playbooks/README.md [hunk 8]
```diff
@@ -49,11 +49,11 @@ Before running the playbook, make sure the following settings are correct, and o
 
 The playbook can be run using:
 ```
-ansible-playbook -i hosts -K install.yml
+ansible-playbook -i hosts -b install.yml
 ```
-The `-K` option will ask for your sudo password, and you have to supply a valid hosts file (here named `hosts`).
+The `-b` option will assume you can become root without a sudo password, and you have to supply a valid hosts file (here named `hosts`).
```
[the remark this answers is not in the document]
Comment (terjekv) [reply ansible/playbooks/README.md:54]: Added, thanks!

File: ansible/playbooks/roles/compatibility_layer/defaults/main.yml (hunk 1 above)
Comment (terjekv) [reply ansible/playbooks/roles/compatibility_layer/defaults/main.yml:18]: Ack!

File: ansible/playbooks/roles/compatibility_layer/defaults/main.yml [hunk 9]
```diff
@@ -9,7 +9,17 @@ cvmfs_publish_transaction: False
...
+
+# How to build the prefix
+prefix_singularity_command: "singularity run -B{{ gentoo_prefix_path }}:{{ gentoo_prefix_path }}"
```
[the remark this answers is not in the document]
Comment (terjekv) [reply ansible/playbooks/roles/compatibility_layer/defaults/main.yml:15]: Thanks, I wasn't sure if Singularity accepted the space. 👍

File: ansible/playbooks/roles/compatibility_layer/defaults/main.yml (hunk 2 above)
Comment (terjekv) [reply ansible/playbooks/roles/compatibility_layer/defaults/main.yml:22]: Much, thanks!

File: ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml (hunk 3 above)
Comment (terjekv) [reply ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml:1]: Yup. My bad.

File: ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    mode: "0755"
+  with_items:
+    - "{{ gentoo_prefix_path }}"
+    - "{{ eessi_log_dir }}"
+  tags:
+    - build_prefix
+
+- name: "Give {{ username_on_host.stdout }} recursive user and group ownership of /cvmfs"
+  file:
+    dest: "/cvmfs"
```
Comment (bedroge) [inline ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml:44]: Shouldn't this be just `{{ gentoo_prefix_path }}`? (and also in the title)

File: ansible/playbooks/roles/compatibility_layer/tasks/install_packages.yml [hunk 11]
```diff
@@ -1,15 +1,29 @@
...
+
+- name: "Get the username running the deployment (not root)"
+  become: no
+  shell: set -o pipefail && who | grep -v root | awk '{print $1}' | uniq
+  register: username_on_host
+  changed_when: True
+
+- name: "Fix permissions after installing as portage/root"
+  become: yes
+  shell: chown -R {{ username_on_host }}:{{ username_on_host }} {{ gentoo_prefix_path }}
```
Comment (bedroge) [inline ansible/playbooks/roles/compatibility_layer/tasks/install_packages.yml:27]: Can this be done with the `file` module and `recurse: yes`?

**Review by bedroge — CHANGES_REQUESTED** [review]

**Comment by bedroge** [comment]
Maybe we should also add a note somewhere that it only works on Centos/RHEL?

File: ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml (hunk 10 above)
Comment (terjekv) [reply ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml:44]: Well... I saw something weird happen with permissions on /cvmfs itself. I haven't been able to reproduce it though.

File: ansible/playbooks/roles/compatibility_layer/tasks/install_packages.yml (hunk 11 above)
Comment (terjekv) [reply ansible/playbooks/roles/compatibility_layer/tasks/install_packages.yml:27]: Yup. I should fix that.

File: ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml (hunk 10 above)
Comment (terjekv) [reply ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml:44]: Right, okay, so we need a become:yes for our overlay due to issues with /proc, which is what causes this follow-up issue.

**Comment by terjekv** [comment]
We should do two things:

1. Check ansible_distribution and fail if it's an unsupported distribution, and
2. Apply some deb/apt support. :)

File: ansible/playbooks/roles/compatibility_layer/tasks/install_packages.yml (hunk 11 above)
Comment (bedroge) [reply ansible/playbooks/roles/compatibility_layer/tasks/install_packages.yml:27]: Trying it with this now:
```
- name: "Fix permissions after installing as portage/root"
  become: yes
  file:
    owner: "{{ username_on_host }}"
    group: "{{ username_on_host }}"
    path: "{{ gentoo_prefix_path }}"
    recurse: true

```

File: ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml (hunk 10 above)
Comment (bedroge) [reply ansible/playbooks/roles/compatibility_layer/tasks/install_prefix.yml:44]: and I'm testing this with the following:
```
- name: "Give {{ username_on_host.stdout }} recursive user and group ownership of {{ gentoo_prefix_path }}"
  file:
    dest: "{{ gentoo_prefix_path }}"
    owner: "{{ username_on_host.stdout }}"
    group: "{{ username_on_host.stdout }}"
    recurse: yes
  tags:
    - build_prefix
```

File: ansible/playbooks/roles/compatibility_layer/defaults/main.yml [hunk 12]
```diff
@@ -1,5 +1,7 @@
 # Defaults file for the compatibility layer role.
 ---
+eessi_overlay_version: "2020.10"
```
Comment (boegel) [inline ansible/playbooks/roles/compatibility_layer/defaults/main.yml:3]: We should bump this to `2020.11` (we won't be rebuilding the `2020.10` compat layer anyway)?

**Review by boegel — CHANGES_REQUESTED** [review]

File: ansible/playbooks/roles/compatibility_layer/defaults/main.yml (hunk 12 above)
Comment (bedroge) [reply ansible/playbooks/roles/compatibility_layer/defaults/main.yml]: Done. Also changed the name of that variable to just `eessi_version`, as overlay means something else in this context.

**Comment by bedroge** [comment]
We also need to update the CI and set `eessi_host_arch` in its `hosts` file.

**Comment by bedroge** [comment]
> We also need to update the CI and set `eessi_host_arch` in its `hosts` file.

Done! Hardcoded it to `x86_64` for now.

**Comment by boegel** [comment]
> > We also need to update the CI and set `eessi_host_arch` in its `hosts` file.
> 
> Done! Hardcoded it to `x86_64` for now.

Why not use `$(uname -m)`? We can also test for `aarch64` on Travis, for example...

**Comment by bedroge** [comment]
> > > We also need to update the CI and set `eessi_host_arch` in its `hosts` file.
> > 
> > 
> > Done! Hardcoded it to `x86_64` for now.
> 
> Why not use `$(uname -m)`? We can also test for `aarch64` on Travis, for example...

That's better, indeed. Changed it.

**Comment by bedroge** [comment]
> We should do two things:
> 
> 1. Check ansible_distribution and fail if it's an unsupported distribution, and
> 2. Apply some deb/apt support. :)

Apparently, there is no Singularity package for Ubuntu 20.04, so we need some ugly checks here. Maybe we should just make a note for now that it only works on Centos (7 and/or 8), and fail if another distro is being used?

**Comment by boegel** [comment]
> > We should do two things:
> > 
> > 1. Check ansible_distribution and fail if it's an unsupported distribution, and
> > 2. Apply some deb/apt support. :)
> 
> Apparently, there is no Singularity package for Ubuntu 20.04 (and maybe some other debian-based distros?), so we need some ugly checks here. Maybe we should just make a note for now that it only works on Centos (7 and/or 8), and fail if another distro is being used?

Yes, agreed. That's good enough for now, I'd say...

I tried this PR + #61, which worked great on an AWS Graviton2 instance to test the build of `2020.11/compat/aarch64`...

**Comment by terjekv** [comment]
Right, CentOS only. Considering the use case (build only), it is probably not a bad idea to have a fairly strict lid on the environment for now.

**Comment by bedroge** [comment]
It now checks the OS of the installation host and fails if it's not Redhat 8, and the snapshot + python version can now be specified in the configuration.

**Review by bedroge — APPROVED** [review]

**Review by peterstol — APPROVED** [review]

**Review by peterstol — APPROVED** [review]

**Review by peterstol — APPROVED** [review]

**Review by peterstol — APPROVED** [review]

**Pull request merged**